subsys/bluetooth/audio/has_client.c

```diff
@@ -727,88 +934,352 @@ static uint8_t features_notify_cb(struct bt_conn *conn, struct bt_gatt_subscribe
+		LOG_DBG("Subscribed already");
+		features_subscribe_cb(inst->conn, 0, &inst->features_sub);
+
+		return 0;
+	}
+
+	inst->features_sub.notify = features_notify_cb;
+	inst->features_sub.subscribe = features_subscribe_cb;
+
+	err = bt_gatt_subscribe(inst->conn, inst->params);
+	if (err != 0) {
+		LOG_ERR("bt_gatt_subscribe err %d", err);
+		inst->features_sub.notify = NULL;
+		inst->params = NULL;
+
+		return err;
+	}
 
-	return bt_gatt_subscribe(inst->conn, &inst->features_subscription);
+	return 0;
+}
+
+static bool is_chrc_notifiable(const struct bt_gatt_subscribe_params *params)
+{
+	return params->value != 0 && params->ccc_handle != 0;
+}
+
+static bool check_ccc_discovery_results(const struct bt_has_client *inst)
+{
+	/* Check whether all the expected CCC dexcriptors have been found */
+	return !!inst->features_sub.value == !!inst->features_sub.ccc_handle &&
+	       !!inst->control_point_sub.value == !!inst->control_point_sub.ccc_handle &&
+	       !!inst->active_index_sub.value == !!inst->active_index_sub.ccc_handle;
+}
+
+static uint8_t ccc_discover_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			       struct bt_gatt_discover_params *params)
+{
+	struct bt_has_client *inst = CONTAINER_OF(params, struct bt_has_client, discover_params);
+
+	if (!attr) {
+		inst->params = NULL;
+
+		if (!check_ccc_discovery_results(inst)) {
+			inst->err = -ENOENT;
+		} else if (is_chrc_notifiable(&inst->features_sub)) {
+			inst->err = features_subscribe(inst);
+		} else {
+			inst->err = features_read(inst);
+		}
+
+		if (inst->err != 0) {
+			inst_state_set(inst, STATE_DISCONNECTING);
+		}
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (bt_uuid_cmp(attr->uuid, BT_UUID_HAS_HEARING_AID_FEATURES) == 0 ||
+	    bt_uuid_cmp(attr->uuid, BT_UUID_HAS_PRESET_CONTROL_POINT) == 0 ||
+	    bt_uuid_cmp(attr->uuid, BT_UUID_HAS_ACTIVE_PRESET_INDEX) == 0) {
+		/* Store the characteristic value UUID to know which characteristic the next CCC
+		 * belongs to.
+		 */
+		(void)memcpy(&inst->uuid, attr->uuid, sizeof(inst->uuid));
+	} else if (bt_uuid_cmp(attr->uuid, BT_UUID_GATT_CHRC) == 0) {
+		/* Clear the characteristic value UUID stored */
+		(void)memset(&inst->uuid, 0, sizeof(inst->uuid));
+	} else if (bt_uuid_cmp(attr->uuid, BT_UUID_GATT_CCC) == 0) {
+		/* Based on characteristic value UUID stored, assign the CCC value handle */
+		if (bt_uuid_cmp(&inst->uuid.uuid, BT_UUID_HAS_HEARING_AID_FEATURES) == 0) {
+			LOG_DBG("features ccc_handle 0x%04x", attr->handle);
+
+			inst->features_sub.ccc_handle = attr->handle;
+		} else if (bt_uuid_cmp(&inst->uuid.uuid, BT_UUID_HAS_PRESET_CONTROL_POINT) == 0) {
+			LOG_DBG("control point ccc_handle 0x%04x", attr->handle);
+
+			inst->control_point_sub.ccc_handle = attr->handle;
+		} else if (bt_uuid_cmp(&inst->uuid.uuid, BT_UUID_HAS_ACTIVE_PRESET_INDEX) == 0) {
+			LOG_DBG("active_index ccc_handle 0x%04x", attr->handle);
+
+			inst->active_index_sub.ccc_handle = attr->handle;
+		}
+	}
+
+	return BT_GATT_ITER_CONTINUE;
```

Comment:
Since most other profiles rely on the auto discover of CCCDs, it's would always be enabled. What's the rationale for not using that?